# Лабораторная работа 7
# + Бонус
## Рекомендательные системы

## Выполнил: __Щеткин Дмитрий Сергеевич__

группа 2.1

In [1]:
import warnings
import pprint
from math import sqrt
import numpy as np

warnings.filterwarnings("ignore", category=FutureWarning)
pp = pprint.PrettyPrinter(indent=4)

In [2]:
# Словарь кинокритиков и выставленных ими оценок
critics={'Lisa Rose': {'Lady in the Water': 2.5, 'Snakes on a Plane': 3.5,
 'Just My Luck': 3.0, 'Superman Returns': 3.5, 'You, Me and Dupree': 2.5,
 'The Night Listener': 3.0},
'Gene Seymour': {'Lady in the Water': 3.0, 'Snakes on a Plane': 3.5,
 'Just My Luck': 1.5, 'Superman Returns': 5.0, 'The Night Listener': 3.0,
 'You, Me and Dupree': 3.5},
'Michael Phillips': {'Lady in the Water': 2.5, 'Snakes on a Plane': 3.0,
 'Superman Returns': 3.5, 'The Night Listener': 4.0},
'Claudia Puig': {'Snakes on a Plane': 3.5, 'Just My Luck': 3.0,
 'The Night Listener': 4.5, 'Superman Returns': 4.0,
 'You, Me and Dupree': 2.5},
'Mick LaSalle': {'Lady in the Water': 3.0, 'Snakes on a Plane': 4.0,
 'Just My Luck': 2.0, 'Superman Returns': 3.0, 'The Night Listener': 3.0,
 'You, Me and Dupree': 2.0},
'Jack Matthews': {'Lady in the Water': 3.0, 'Snakes on a Plane': 4.0,
 'The Night Listener': 3.0, 'Superman Returns': 5.0, 'You, Me and Dupree': 3.5},
'Toby': {'Snakes on a Plane':4.5,'You, Me and Dupree':1.0,'Superman Returns':4.0}}

In [3]:
pp.pprint(critics)

{   'Claudia Puig': {   'Just My Luck': 3.0,
                        'Snakes on a Plane': 3.5,
                        'Superman Returns': 4.0,
                        'The Night Listener': 4.5,
                        'You, Me and Dupree': 2.5},
    'Gene Seymour': {   'Just My Luck': 1.5,
                        'Lady in the Water': 3.0,
                        'Snakes on a Plane': 3.5,
                        'Superman Returns': 5.0,
                        'The Night Listener': 3.0,
                        'You, Me and Dupree': 3.5},
    'Jack Matthews': {   'Lady in the Water': 3.0,
                         'Snakes on a Plane': 4.0,
                         'Superman Returns': 5.0,
                         'The Night Listener': 3.0,
                         'You, Me and Dupree': 3.5},
    'Lisa Rose': {   'Just My Luck': 3.0,
                     'Lady in the Water': 2.5,
                     'Snakes on a Plane': 3.5,
                     'Superman Returns': 3.5,
                 

**Упражнение.** Выведите оценки, которые выставил Toby

In [4]:
pp.pprint(critics['Toby'])

{'Snakes on a Plane': 4.5, 'Superman Returns': 4.0, 'You, Me and Dupree': 1.0}


**Упражнение.** Выведите оценку критика Lisa Rose для фильма Lady in the Water

In [5]:
pp.pprint(critics['Lisa Rose']['Lady in the Water'])

2.5


**Упражнение.**  Реализуйте функцию для вычисления расстояния Евклида

In [6]:
def compute_euclid(x, y):
    squared_diffs = [(xi - yi)**2 for xi, yi in zip(x, y)]
    result = sqrt(sum(squared_diffs))
    return result

In [7]:
compute_euclid([1, 1], [2, 3])
# должно получиться:  2.23606797749979

2.23606797749979

In [8]:
compute_euclid([1, 1], [8, 10])
# должно получиться: 11.40175425099138

11.40175425099138

In [9]:
1 / (1 + compute_euclid([1, 1], [2, 3]))
# Должно получиться 0.309016

0.3090169943749474

In [10]:
1 / (1 + compute_euclid([1, 1], [8, 10]))
# Должно получиться: 0.08063375388365411

0.08063375388365411

In [11]:
%%capture
!wget https://www.dropbox.com/s/udwga4a1tohd99b/recommendations.py

In [12]:
from recommendations import *

In [13]:
sim_distance(critics, 'Lisa Rose', 'Toby')

0.2222222222222222

**Упражнение.**  Реализуйте функцию для вычисления расстояния sim_distance для двух критиков: "Mick LaSalle" и "Gene Seymour"

In [14]:
sim_distance(critics, 'Mick LaSalle', 'Gene Seymour')
# Должно получиться 0.12903

0.12903225806451613

In [15]:
def pearson_corr(x, y):
    n = len(x)
    mean_x = np.mean(x)
    mean_y = np.mean(y)

    numerator = sum((xi - mean_x)*(yi - mean_y) for xi, yi in zip(x, y))

    denominator = np.sqrt(sum((xi - mean_x)**2 for xi in x)) * np.sqrt(sum((yi - mean_y)**2 for yi in y))

    result = numerator / denominator
    return result

In [16]:
x = []
y = []
for film in critics['Lisa Rose']:
    x.append(critics['Lisa Rose'][film])
    y.append(critics['Gene Seymour'][film])

print(pearson_corr(x, y)) # должно быть: 0.39605901719066977

0.39605901719066977


In [17]:
sim_pearson(critics, 'Lisa Rose', 'Gene Seymour')

np.float64(0.3960590171906697)

**Упражнение.**  Найдите в модуле scipy как вычисляется корреляция Пирсона и получите аналогичное значение (0.3960590171906697) как и с помощью вручную написанной функции.

In [18]:
from scipy.stats import pearsonr

common_movies = set(critics['Lisa Rose'].keys()) & set(critics['Gene Seymour'].keys())
lisa_ratings = [critics['Lisa Rose'][movie] for movie in common_movies]
gene_ratings = [critics['Gene Seymour'][movie] for movie in common_movies]

corr_coefficient, _ = pearsonr(lisa_ratings, gene_ratings)
print(corr_coefficient)

0.39605901719066977


In [19]:
topMatches(critics, 'Toby', n=3, similarity=sim_pearson)

[(np.float64(0.9912407071619302), 'Lisa Rose'),
 (np.float64(0.924473451641905), 'Mick LaSalle'),
 (np.float64(0.8934051474415643), 'Claudia Puig')]

**Упражнение.** Вычислите топ-3 критиков похожих на Toby по евклидову расстоянию (sim_distance):

In [20]:
similarities = []
for critic in critics.keys():
    if critic != 'Toby':
        similarities.append((sim_distance(critics, 'Toby', critic), critic))

top_3_matches = sorted(similarities, reverse=True)[:3]

for match in top_3_matches:
    print(match)

(0.3076923076923077, 'Mick LaSalle')
(0.2857142857142857, 'Michael Phillips')
(0.23529411764705882, 'Claudia Puig')


**Упражнение.** Вычислите рекомендации для ```Toby``` с использованием ```sim_distance``` в качестве функции-критерия для сравнения:

In [21]:
getRecommendations(critics, 'Toby', sim_distance)

[(3.5002478401415877, 'The Night Listener'),
 (2.7561242939959363, 'Lady in the Water'),
 (2.461988486074374, 'Just My Luck')]

In [22]:
itemsim = calculateSimilarItems(critics, n = 5)

In [23]:
pp.pprint(itemsim)

{   'Just My Luck': [   (0.2222222222222222, 'Lady in the Water'),
                        (0.18181818181818182, 'You, Me and Dupree'),
                        (0.15384615384615385, 'The Night Listener'),
                        (0.10526315789473684, 'Snakes on a Plane'),
                        (0.06451612903225806, 'Superman Returns')],
    'Lady in the Water': [   (0.4, 'You, Me and Dupree'),
                             (0.2857142857142857, 'The Night Listener'),
                             (0.2222222222222222, 'Snakes on a Plane'),
                             (0.2222222222222222, 'Just My Luck'),
                             (0.09090909090909091, 'Superman Returns')],
    'Snakes on a Plane': [   (0.2222222222222222, 'Lady in the Water'),
                             (0.18181818181818182, 'The Night Listener'),
                             (0.16666666666666666, 'Superman Returns'),
                             (0.10526315789473684, 'Just My Luck'),
                             (

**Упражнение.** Получите список рекомендованных фильмов, используя данные о схожести ```itemsim``` и критика ```Toby```:

In [24]:
getRecommendedItems(critics, itemsim, 'Toby')

[(3.182634730538922, 'The Night Listener'),
 (2.5983318700614575, 'Just My Luck'),
 (2.4730878186968837, 'Lady in the Water')]

**Упражнение.** Получите с помощью функции `getRecommendations` для критика Toby с помощью функции `sim_pearson` список рекомендаций:

In [25]:
getRecommendations(critics, 'Toby', sim_pearson)

[(np.float64(3.347789526713101), 'The Night Listener'),
 (np.float64(2.832549918264161), 'Lady in the Water'),
 (np.float64(2.5309807037655645), 'Just My Luck')]

## Бонус* : Рекомендация на данных MovieLens

In [26]:
%%capture
!wget https://www.dropbox.com/sh/ob1qyfzqygkc6ph/AADg2G0By0EuRZCLR5gSAmP7a

In [27]:
%%capture
!unzip AADg2G0By0EuRZCLR5gSAmP7a -d data

In [28]:
preferences = loadMovieLens()

In [29]:
pp.pprint(preferences['87'])

{   '2001: A Space Odyssey (1968)': 5.0,
    'Ace Ventura: Pet Detective (1994)': 4.0,
    'Addams Family Values (1993)': 2.0,
    'Addicted to Love (1997)': 4.0,
    'Adventures of Priscilla, Queen of the Desert, The (1994)': 3.0,
    'Adventures of Robin Hood, The (1938)': 5.0,
    'Air Force One (1997)': 3.0,
    'Air Up There, The (1994)': 3.0,
    'Alien (1979)': 4.0,
    'American President, The (1995)': 5.0,
    'Annie Hall (1977)': 4.0,
    'Apocalypse Now (1979)': 4.0,
    'Babe (1995)': 5.0,
    'Baby-Sitters Club, The (1995)': 2.0,
    'Back to the Future (1985)': 5.0,
    'Bad Boys (1995)': 4.0,
    'Bananas (1971)': 5.0,
    'Barcelona (1994)': 3.0,
    'Batman & Robin (1997)': 4.0,
    'Batman (1989)': 3.0,
    'Batman Returns (1992)': 3.0,
    'Big Green, The (1995)': 3.0,
    'Big Squeeze, The (1996)': 2.0,
    'Birdcage, The (1996)': 4.0,
    'Blade Runner (1982)': 4.0,
    'Blues Brothers, The (1980)': 5.0,
    'Boomerang (1992)': 3.0,
    'Boot, Das (1981)': 4.0,
   

In [30]:
getRecommendations(preferences, '87')[0:30]

/content/recommendations.py:58: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  r = pearsonr(scores1, scores2)[0]


[(np.float64(5.0), 'They Made Me a Criminal (1939)'),
 (np.float64(5.0), 'Star Kid (1997)'),
 (np.float64(5.0), 'Santa with Muscles (1996)'),
 (np.float64(5.0), 'Saint of Fort Washington, The (1993)'),
 (np.float64(5.0), 'Marlene Dietrich: Shadow and Light (1996) '),
 (np.float64(5.0), 'Great Day in Harlem, A (1994)'),
 (np.float64(5.0), 'Entertaining Angels: The Dorothy Day Story (1996)'),
 (np.float64(5.0), 'Boys, Les (1997)'),
 (np.float64(4.898844431289231), 'Legal Deceit (1997)'),
 (np.float64(4.815019082242706), 'Letter From Death Row, A (1998)'),
 (np.float64(4.732108298394142), 'Hearts and Minds (1996)'),
 (np.float64(4.6962444664908665), 'Pather Panchali (1955)'),
 (np.float64(4.652397061026759), 'Lamerica (1994)'),
 (np.float64(4.538723693474812), 'Leading Man, The (1996)'),
 (np.float64(4.535081339106105), 'Mrs. Dalloway (1997)'),
 (np.float64(4.532337612572983), 'Innocents, The (1961)'),
 (np.float64(4.527998574747079), 'Casablanca (1942)'),
 (np.float64(4.5102701497198625)

In [31]:
item_similarities = calculateSimilarItems(preferences, n = 50)

100 / 1664
200 / 1664
300 / 1664
400 / 1664
500 / 1664
600 / 1664
700 / 1664
800 / 1664
900 / 1664
1000 / 1664
1100 / 1664
1200 / 1664
1300 / 1664
1400 / 1664
1500 / 1664
1600 / 1664


In [32]:
item_similarities["What's Eating Gilbert Grape (1993)"]

[(1.0, 'Yankee Zulu (1994)'),
 (1.0, 'Woman in Question, The (1950)'),
 (1.0, 'Witness (1985)'),
 (1.0, "Wend Kuuni (God's Gift) (1982)"),
 (1.0, 'Vie est belle, La (Life is Rosey) (1987)'),
 (1.0, 'Vermont Is For Lovers (1992)'),
 (1.0, 'Van, The (1996)'),
 (1.0, 'Two or Three Things I Know About Her (1966)'),
 (1.0, 'Two Much (1996)'),
 (1.0, 'Two Deaths (1995)'),
 (1.0, 'Two Bits (1995)'),
 (1.0, 'Touki Bouki (Journey of the Hyena) (1973)'),
 (1.0, 'Total Eclipse (1995)'),
 (1.0, 'To Have, or Not (1995)'),
 (1.0, 'To Cross the Rubicon (1991)'),
 (1.0, 'Tigrero: A Film That Was Never Made (1994)'),
 (1.0, 'The Courtyard (1995)'),
 (1.0, 'Terror in a Texas Town (1958)'),
 (1.0, 'Talking About Sex (1994)'),
 (1.0, 'T-Men (1947)'),
 (1.0, 'Symphonie pastorale, La (1946)'),
 (1.0, 'Swept from the Sea (1997)'),
 (1.0, 'Stranger in the House (1997)'),
 (1.0, 'Stonewall (1995)'),
 (1.0, 'Squeeze (1996)'),
 (1.0, 'Spirits of the Dead (Tre passi nel delirio) (1968)'),
 (1.0, 'Somebody to Love

In [33]:
getRecommendedItems(preferences, item_similarities, '87')[0:30]

[(5.0, "What's Eating Gilbert Grape (1993)"),
 (5.0, 'Vertigo (1958)'),
 (5.0, 'Usual Suspects, The (1995)'),
 (5.0, 'Toy Story (1995)'),
 (5.0, 'Titanic (1997)'),
 (5.0, 'Sword in the Stone, The (1963)'),
 (5.0, 'Stand by Me (1986)'),
 (5.0, 'Sling Blade (1996)'),
 (5.0, 'Silence of the Lambs, The (1991)'),
 (5.0, 'Shining, The (1980)'),
 (5.0, 'Shine (1996)'),
 (5.0, 'Sense and Sensibility (1995)'),
 (5.0, 'Scream (1996)'),
 (5.0, 'Rumble in the Bronx (1995)'),
 (5.0, 'Rock, The (1996)'),
 (5.0, 'Robin Hood: Prince of Thieves (1991)'),
 (5.0, 'Reservoir Dogs (1992)'),
 (5.0, 'Police Story 4: Project S (Chao ji ji hua) (1993)'),
 (5.0, 'House of the Spirits, The (1993)'),
 (5.0, 'Fresh (1994)'),
 (5.0, 'Denise Calls Up (1995)'),
 (5.0, 'Day the Sun Turned Cold, The (Tianguo niezi) (1994)'),
 (5.0, 'Before the Rain (Pred dozhdot) (1994)'),
 (5.0, 'Assignment, The (1997)'),
 (5.0, '1-900 (1994)'),
 (4.875, "Ed's Next Move (1996)"),
 (4.833333333333333, 'Anna (1996)'),
 (4.8, 'Dark City 

In [34]:
top_films = [(rating, films) for films, rating in preferences['87'].items() if rating == 5.0]
pp.pprint(top_films)

[   (5.0, 'Jurassic Park (1993)'),
    (5.0, 'Indiana Jones and the Last Crusade (1989)'),
    (5.0, 'Good, The Bad and The Ugly, The (1966)'),
    (5.0, 'Dead Poets Society (1989)'),
    (5.0, 'M*A*S*H (1970)'),
    (5.0, '2001: A Space Odyssey (1968)'),
    (5.0, 'Four Weddings and a Funeral (1994)'),
    (5.0, 'Prefontaine (1997)'),
    (5.0, 'This Is Spinal Tap (1984)'),
    (5.0, 'Cool Hand Luke (1967)'),
    (5.0, 'True Lies (1994)'),
    (5.0, 'Sleepless in Seattle (1993)'),
    (5.0, 'Young Frankenstein (1974)'),
    (5.0, 'Fugitive, The (1993)'),
    (5.0, 'In the Line of Fire (1993)'),
    (5.0, 'Speed (1994)'),
    (5.0, 'Terminator 2: Judgment Day (1991)'),
    (5.0, 'Shawshank Redemption, The (1994)'),
    (5.0, 'Clear and Present Danger (1994)'),
    (5.0, 'Dances with Wolves (1990)'),
    (5.0, 'Adventures of Robin Hood, The (1938)'),
    (5.0, 'Blues Brothers, The (1980)'),
    (5.0, 'Babe (1995)'),
    (5.0, 'Top Gun (1986)'),
    (5.0, 'Get Shorty (1995)'),
    (5.0, 

При визуальном сравнении видно, что практически нет общих фильмов между первым набором рекомендованных фильмов и вторым набором фильмов, оценённых пользователем на 5.0.